System/Posix/Fcntl.hsc

```diff
@@ -21,18 +21,30 @@ module System.Posix.Fcntl (
     -- * File allocation
     Advice(..), fileAdvise,
     fileAllocate,
+    -- * File caching
+    fileGetCaching,
+    fileSetCaching,
   ) where
 
-#if HAVE_POSIX_FALLOCATE || HAVE_POSIX_FADVISE
 import Foreign.C
-#endif
 import System.Posix.Types
 
 #if !HAVE_POSIX_FALLOCATE
 import System.IO.Error ( ioeSetLocation )
 import GHC.IO.Exception ( unsupportedOperation )
 #endif
 
+#ifndef darwin_HOST_OS
```

Comment:
Why is this check performed on darwin?

ghc-internals does [this](https://hackage.haskell.org/package/ghc-internal-9.1201.0/docs/src/GHC.Internal.System.Posix.Internals.html#c_fcntl_read):

```hs
#if !defined(mingw32_HOST_OS) && !defined(javascript_HOST_ARCH)
```

This already suggests it doesn't work with the JS backend. Further, I'd like to avoid platform ifdefs in unix. We can use `configure.ac` to check, e.g.:

* https://github.com/filebench/filebench/blob/22620e602cbbebad90c0bd041896ebccf70dbf5f/configure.ac#L414-L432
* or https://browse.dgit.debian.org/evolution-data-server.git/plain/libdb/dist/configure.ac?id=1a1b76736a541165dc607de9ccf20d3ad7017990

In the latter case, I think they check if it has an effect... we probably don't need to go that far.

Reply:
Makes sense -- I've modelled it after `filebench`'s approach, but made it fit the style of the `configure.ac` file

I'm not very familiar with `autoconf`, so let me know if anything is off

Reply:
The darwin check was performed like that because we can't set `O_DIRECT` on OSX -- we have to set the `F_NOCACHE` flag instead. But the check is rewritten now to case on `HAVE_O_DIRECT` and `HAVE_F_NOCACHE`

Reply:
( I actually forgot to change the platform ifdef in the imports, but it should now be removed there as well )